propsigner/propsigner_test.go

```diff
@@ -0,0 +1,157 @@
+package propsigner
+
+import (
+	"bytes"
+	"context"
+	"testing"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/go-state-types/crypto"
+	"github.com/filecoin-project/specs-actors/actors/builtin/market"
+	"github.com/ipfs/go-cid"
+	libwal "github.com/jsign/go-filsigner/wallet"
+	"github.com/libp2p/go-libp2p-core/peer"
+	swarmt "github.com/libp2p/go-libp2p-swarm/testing"
+	bhost "github.com/libp2p/go-libp2p/p2p/host/basic"
+	"github.com/stretchr/testify/require"
+	"github.com/textileio/go-auctions-client/localwallet"
+)
+
+var (
+	walletKeys = []string{
+		// Secp256k1 exported private key in Lotus format.
+		"7b2254797065223a22736563703235366b31222c22507269766174654b6579223a226b35507976337148327349586343595a58594f5775453149326e32554539436861556b6c4e36695a5763453d227d", // nolint:lll
+		// BLS exported private key in Lotus format.
+		"7b2254797065223a22626c73222c22507269766174654b6579223a226862702f794666527439514c43716b6d566171415752436f50556777314b776971716e73684e49704e57513d227d", // nolint:lll
+	}
+)
+
+type testCase struct {
+	name      string
+	proposal  market.DealProposal
+	authToken string
+	err       error
+}
+
+func TestProposalSigning(t *testing.T) {
+	t.Parallel()
+
+	authToken := "veryhardtokentoguess"
+	wallet, err := localwallet.New(walletKeys)
+	require.NoError(t, err)
+
+	testCases := []testCase{
+		{
+			name:      "success secp256k1",
+			proposal:  correctProposalSecp256k1(t),
+			authToken: authToken,
+			err:       nil,
+		},
+		{
+			name:      "success bls",
+			proposal:  correctProposalBLS(t),
+			authToken: authToken,
+			err:       nil,
+		},
+		{
+			name:      "invalid auth token",
+			proposal:  correctProposalSecp256k1(t),
+			authToken: "wrongToken",
+			err:       errInvalidAuthToken,
+		},
+		{
+			name:      "invalid auth token",
+			proposal:  proposalWithUnknownAddress(t),
+			authToken: authToken,
+			err:       errWalletMissingKeys,
+		},
+	}
+
+	for _, test := range testCases {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			// Remote wallet libp2p host.
+			h1 := bhost.New(swarmt.GenSwarm(t, ctx))
+			err := NewDealSignerService(h1, authToken, wallet)
+			require.NoError(t, err)
+
+			// Client (dealerd) libp2p2 host.
+			h2 := bhost.New(swarmt.GenSwarm(t, ctx))
+			err = h2.Connect(ctx, peer.AddrInfo{ID: h1.ID(), Addrs: h1.Addrs()})
+			require.NoError(t, err)
+
+			sig, err := RequestSignatureV1(ctx, h2, test.authToken, test.proposal, h1.ID())
+			if test.err != nil {
+				require.Error(t, err)
+				require.Contains(t, err.Error(), test.err.Error())
+			} else {
+				require.NoError(t, err)
+				requireValidSignature(t, test.proposal, sig)
```

Comment:
In success cases, just be paranoid and also check that the returned signature is valid from a client point of view.

`dealerd` will always check the signatures before doing the proposal to the miner, just to avoid buggy clients sending wrong signatures (maybe if they do their own implementation or some bug in the code).